File: internal/application/circuit_breaker.go
package application

import "time"

type CircuitBreaker struct {
	failures    int
	threshold   int
	resetAfter  time.Duration
	lastFailure time.Time
}

func NewCircuitBreaker(threshold int, reset time.Duration) *CircuitBreaker {
	return &CircuitBreaker{
		threshold:  threshold,
		resetAfter: reset,
	}
}

func (c *CircuitBreaker) Allow() bool {

	if c.failures < c.threshold {
		return true
	}

	if time.Since(c.lastFailure) > c.resetAfter {
		c.failures = 0
		return true
	}

	return false
}

func (c *CircuitBreaker) Fail() {
	c.failures++
	c.lastFailure = time.Now()
}

func (c *CircuitBreaker) Success() {
	c.failures = 0
}

File: application/circuit_breaker.ts
type State = "closed" | "open" | "half-open";

export interface CircuitBreaker {
    allow(): boolean;
    fail(): void;
    success(): void;
}

export function createCircuitBreaker(
    threshold: number,
    resetAfter: number,
): CircuitBreaker {
    let state: State = "closed";
    let failures = 0;
    let lastFailure = 0;

    return {
        allow() {
            switch (state) {
                case "closed":
                    return true;
                case "open":
                    if (Date.now() - lastFailure > resetAfter) {
                        state = "half-open";
                        return true;
                    }
                    return false;
                case "half-open":
                    return false;
            }
        },

        fail() {
            failures++;
            lastFailure = Date.now();
            if (failures >= threshold) {
                state = "open";
            }
        },

        success() {
            failures = 0;
            state = "closed";
        },
    };
}